src/scene/vcModel.cpp

```diff
@@ -758,10 +689,33 @@ udDouble3 vcModel::GetLocalSpacePivot()
 
 udDouble4x4 vcModel::GetWorldSpaceMatrix()
 {
-  return m_sceneMatrix;
+  return udDouble4x4::translation(m_pivot) * udDouble4x4::rotationQuat(m_orientationCurrent, m_positionCurrent) * udDouble4x4::scaleUniform(m_scaleCurrent.x) * udDouble4x4::translation(-m_pivot);
 }
 
 vcGizmoAllowedControls vcModel::GetAllowedControls()
 {
   return vcGAC_AllUniform;
 }
+
+void vcModel::SetTransforms(const udDouble4x4 &matrix)
+{
+  (udDouble4x4::translation(-m_pivot) * matrix * udDouble4x4::translation(m_pivot)).extractTransforms(m_positionCurrent, m_scaleCurrent, m_orientationCurrent);
+}
+
+void vcModel::SaveTransformsToNode(vcProject * pProject, const udGeoZone &defaultZone)
+{
+  if (m_pCurrentZone != nullptr)
+    vcProject_UpdateNodeGeometryFromCartesian(pProject, m_pNode, *m_pCurrentZone, vdkPGT_Point, &m_positionCurrent, 1);
+  else if (m_pPreferredProjection != nullptr)
+    vcProject_UpdateNodeGeometryFromCartesian(pProject, m_pNode, *m_pPreferredProjection, vdkPGT_Point, &m_positionCurrent, 1);
+  else
+    vcProject_UpdateNodeGeometryFromCartesian(pProject, m_pNode, defaultZone, vdkPGT_Point, &m_positionCurrent, 1);
```

Comment:
I would recommend using something like:
```c++
const udGeoZone *pGeoZone = &defaultZone;
if (m_pCurrentZone != nullptr)
  pGeoZone = m_pCurrentZone;
else if (m_pPreferredProjection != nullptr)
  pGeoZone = m_pPreferredProjection;

vcProject_UpdateNodeGeometryFromCartesion(pProject, m_pNode, *pGeoZone, vdkPGT_Point, &m_positionCurrent, 1);
```
If any of the parameters need to change, or should change, this ensures that bugs aren't accidentally introduced because one of them wasn't updated. Additionally, it makes it abundantly clear to the reader exactly what changes.

Reply:
I like it